File: MojPrijevoz/MojPrijevoz.Model/Requests/FareOffer/FareOfferInsertRequest.cs
using MojPrijevoz.Model.Dtos.Nominatim;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using MojPrijevoz.Model.Requests.StopPoint;

namespace MojPrijevoz.Model.Requests.FareOffer;

public class FareOfferInsertRequest : IValidatableObject {
    [Required] public int OriginCityId { get; set; }
    [Required] public NominatimCityDto DestinationCity { get; set; } = null!;
    [Required][Range(1, int.MaxValue)] public float Length { get; set; }
    [Required][Range(1, int.MaxValue)] public float Duration { get; set; }

    [NotMapped] public int PassengerId { get; set; }
    [NotMapped] public int FareDataId { get; set; }
    [NotMapped] public float Price { get; set; }
    [NotMapped] public int UserVehicleId { get; set; }

    [Required] public ICollection<FareOfferDriverPriceDto> DriversPrices { get; set; } = null!;
    [Required] public IReadOnlyList<StopPointInsertRequest> StopPoints { get; set; } = null!;

    [Required] public DateTime FareDateTime { get; set; }

    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext) {
        if (FareDateTime < DateTime.Now) {
            yield return new ValidationResult("Vrijeme vožnje ne može biti u prošlosti!", [nameof(FareDateTime)]);
        }
    }
}

File: MojPrijevozFront/moj_prijevoz/lib/resources/requests/fare_offer/fare_offer_insert_request.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:moj_prijevoz/resources/dtos/fare_offer/fare_offer_driver_price_dto.dart';
import 'package:moj_prijevoz/resources/dtos/nominatim/nominatim_city_dto.dart';
import 'package:moj_prijevoz/resources/dtos/stop_point/stop_point_dto.dart';
import 'package:moj_prijevoz/utils/json_parser.dart';

part 'fare_offer_insert_request.g.dart';

@JsonSerializable(explicitToJson: true)
class FareOfferInsertRequest extends JsonRequest {
  final int originCityId;
  final NominatimCityDto destinationCity;
  final String destinationName;
  final double length;
  final double duration;
  final List<FareOfferDriverPriceDto> driversPrices;
  final List<StopPointDto>? stopPoints;
  final DateTime fareDateTime;

  FareOfferInsertRequest({
    required this.originCityId,
    required this.destinationCity,
    required this.length,
    required this.duration,
    required this.driversPrices,
    this.stopPoints,
    required this.fareDateTime,
    required this.destinationName,
  });

  @override
  Map<String, dynamic> toJson() => _$FareOfferInsertRequestToJson(this);

  factory FareOfferInsertRequest.fromJson(Map<String, dynamic> json) =>
      _$FareOfferInsertRequestFromJson(json);
}
